sys/dev/usb/usb_generic.c

```diff
@@ -1223,54 +1203,175 @@ ugen_fs_copy_in(struct usb_fifo *f, uint8_t ep_index)
+{
+	void *uptr;
+#ifdef COMPAT_FREEBSD32
+	struct usb_fs_endpoint32 fs_ep32;
+	uint32_t *uptr32;
+#endif
+	int error;
+
+	switch(f->fs_ep_sz) {
+	case sizeof(struct usb_fs_endpoint):
+		error = copyin(ugen_fs_ep_uptr(f, ep_index), fs_ep,
+		    f->fs_ep_sz);
+		uptr = fs_ep->ppBuffer;
+		if (error != 0)
+			return (error);
+		break;
+
+#ifdef COMPAT_FREEBSD32
+	case sizeof(struct usb_fs_endpoint32):
+		error = copyin(ugen_fs_ep_uptr(f, ep_index), &fs_ep32,
+		    f->fs_ep_sz);
+		if (error != 0)
+			return (error);
+		uptr32 = PTRIN(fs_ep32.ppBuffer);
+		PTRIN_CP(fs_ep32, *fs_ep, pLength);
+		CP(fs_ep32, *fs_ep, nFrames);
+		CP(fs_ep32, *fs_ep, aFrames);
+		CP(fs_ep32, *fs_ep, flags);
+		CP(fs_ep32, *fs_ep, timeout);
+		CP(fs_ep32, *fs_ep, isoc_time_complete);
+		CP(fs_ep32, *fs_ep, status);
+		break;
+#endif
+	default:
+		panic("%s: called with unknown size %#zx", __func__, f->fs_ep_sz);
+	}
+
+	if (!copy_arrays || fs_ep->nFrames == 0) {
+		fs_ep->ppBuffer = NULL;
+		return (0);
+	}
+
+	/*
+	 * Avoid userspace DoS.  16 is selected because it's the
+	 * largest number any in-tree driver uses (most use 0-2).
+	 * It might be that MAXPHYS/sizeof(void *) would be better.
+	 */
+	if (fs_ep->nFrames > 16)
+		return (EINVAL);
+
+	switch(f->fs_ep_sz) {
+	case sizeof(struct usb_fs_endpoint):
+		fs_ep->ppBuffer = malloc(sizeof(*fs_ep->ppBuffer) *
+		    fs_ep->nFrames, M_USB, M_WAITOK);
+		error = copyin(uptr, fs_ep->ppBuffer,
+		    sizeof(*fs_ep->ppBuffer) * fs_ep->nFrames);
+		break;
+
+#ifdef COMPAT_FREEBSD32
+	case sizeof(struct usb_fs_endpoint32):
+		fs_ep->ppBuffer = malloc(sizeof(*fs_ep->ppBuffer) *
+		    fs_ep->nFrames, M_USB, M_WAITOK);
+		for (uint32_t i = 0; i < fs_ep->nFrames; i++) {
+			fs_ep->ppBuffer[i] = PTRIN(fuword32(uptr32 + i));
+			if ((intptr_t)fs_ep->ppBuffer[i] == -1) {
+			    error = EFAULT;
+			    break;
+			}
+		}
+		break;
+#endif
+	}
+
+	if (error != 0)
+		free(fs_ep->ppBuffer, M_USB);
+	return (error);
+}
+
 static int
-ugen_fs_copy_out_cancelled(struct usb_fs_endpoint *fs_ep_uptr)
+ugen_fs_update(const struct usb_fs_endpoint *fs_ep,
+    struct usb_fifo *f, uint8_t ep_index)
+{
+	union {
+		struct usb_fs_endpoint *fs_ep_uptr;
+#ifdef COMPAT_FREEBSD32
+		struct usb_fs_endpoint32 *fs_ep_uptr32;
+#endif
+	} u;
+	uint32_t *aFrames_uptr;
+	uint16_t *isoc_time_complete_uptr;
+	int *status_uptr;
+
+	switch (f->fs_ep_sz) {
+	case sizeof(struct usb_fs_endpoint):
+		u.fs_ep_uptr = ugen_fs_ep_uptr(f, ep_index);
+		aFrames_uptr = &u.fs_ep_uptr->aFrames;
+		isoc_time_complete_uptr = &u.fs_ep_uptr->isoc_time_complete;
+		status_uptr = &u.fs_ep_uptr->status;
+		break;
+#ifdef COMPAT_FREEBSD32
+	case sizeof(struct usb_fs_endpoint32):
+		u.fs_ep_uptr32 = (struct usb_fs_endpoint32 *)
+		    ugen_fs_ep_uptr(f, ep_index);
+		aFrames_uptr = &u.fs_ep_uptr32->aFrames;
+		isoc_time_complete_uptr = &u.fs_ep_uptr32->isoc_time_complete;
+		status_uptr = &u.fs_ep_uptr32->status;
+		break;
+#endif
+	}
```

Comment:
default:
break;

Reply:
If there's going to be a `default:` here it must be a panic.